_posts/2023-04-19-more-optimizations.md

```diff
@@ -0,0 +1,1421 @@
+
+```
+    {test,bs_start_match3,{f,1},1,[{x,0}],{x,1}}.
+    {bs_get_position,{x,1},{x,0},2}.
+    {test,bs_get_integer2,
+          {f,2},
+          2,
+          [{x,1},
+           {integer,8},
+           1,
+           {field_flags,[unsigned,big]}],
+          {x,2}}.
+    {test,bs_get_integer2,
+          {f,2},
+          3,
+          [{x,1},
+           {integer,24},
+           1,
+           {field_flags,[unsigned,big]}],
+          {x,3}}.
+    {test,bs_test_tail2,{f,2},[{x,1},0]}.
+    {bs_create_bin,{f,0},
+                   0,4,1,
+                   {x,0},
+                   {list,[{atom,integer},
+                          1,1,nil,
+                          {tr,{x,3},{t_integer,{0,16777215}}},
+                          {integer,24},
+                          {atom,integer},
+                          2,1,nil,
+                          {tr,{x,2},{t_integer,{0,255}}},
+                          {integer,8}]}}.
+    return.
+```
+
+Let's walk through the code. The first instruction sets up a [match
+context](https://www.erlang.org/doc/efficiency_guide/binaryhandling.html#how-binaries-are-implemented):
+
+```
+    {test,bs_start_match3,{f,1},1,[{x,0}],{x,1}}.
+```
+
+A match context holds several pieces of information needed for
+matching a binary.
+
+The next instruction saves information that will be needed if matching
+of the binary fails for some reason:
+
+```
+    {bs_get_position,{x,1},{x,0},2}.
+```
+
+The next two instructions match out two segments as integers (comments added by me):
+
+```
+    {test,bs_get_integer2,
+          {f,2},          % Failure label
+          2,              % Number of live X registers (needed for GC)
+          [{x,1},         % Match context register
+           {integer,8},   % Size of segment in units
+           1,             % Unit value
+           {field_flags,[unsigned,big]}],
+          {x,2}}.         % Destination register
+    {test,bs_get_integer2,
+          {f,2},
+          3,
+          [{x,1},
+           {integer,24},
+           1,
+           {field_flags,unsigned,big]}],
+          {x,3}}.
+```
+
+The next instruction makes sure that the end of the binary has now been
+reached:
+
+```
+    {test,bs_test_tail2,{f,2},[{x,1},0]}.
+```
+
+The next instruction creates the binary with the segments swapped:
+
+```
+    {bs_create_bin,{f,0},
+                   0,4,1,
+                   {x,0},
+                   {list,[{atom,integer},
+                          1,1,nil,
+                          {tr,{x,3},{t_integer,{0,16777215}}},
+                          {integer,24},
+                          {atom,integer},
+                          2,1,nil,
+                          {tr,{x,2},{t_integer,{0,255}}},
+                          {integer,8}]}}.
+```
+
+Before OTP 25, creation of binaries was done using multiple
+instructions, similar to how binary matching is still done in OTP
+25. The reason for the new instruction is to be able to provide
```

Comment:
"new instruction" is a dangling pointer :) It's not clear to me which instruction that is.

Reply:
Will clarify.